### ISSUE TITLE
Pod .spec.schedulerName set as part of deserialization

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#93067]
/sig scheduling
/sig api-machinery

Pod .spec.schedulerName is set as part of deserialization

https://github.com/kubernetes/kubernetes/blob/5e8ccda71cc676e070cd0f5d7b6990b8ebc24c1f/pkg/apis/core/v1/defaults.go#L219-L221

This is business logic and, as such, should live in a create registry function.
The problem with this being part of deserialization is that external webhooks can't detect if it was user intention to use the default scheduler or not.

**Comment by alculquicondor** [comment]
@dbsmith @liggitt for thoughts

@Huang-Wei for visibility

**Comment by lavalamp** [comment]
Since we only have one pod version, I think it's OK to move this from the perspective of following our api conventions, but with one caveat:

If anyone is currently trying to manage this field via webhook they might be surprised if this field suddenly starts coming in blank. I think that's the primary concern. This would actually make their life much easier overall, but it would break them until they make a change.

Should it go in the registry (completely post-webhooks) or as an admission plugin (runs with mutating webhooks)?

**Comment by alculquicondor** [comment]
do admission plugins get the updated versions from previous plugins? do they run before or after webhooks?

**Comment by lavalamp** [comment]
Admission plugins are split into mutating and validating and run with the
webhooks. I don't recall if the mutating ones run before or after the
webhooks, @caesarxuchao might.

**Comment by caesarxuchao** [comment]
According to this [list](https://github.com/kubernetes/kubernetes/blob/01f98c4321cd03e15f16d6d4ea42797f2d4bca3c/pkg/kubeapiserver/options/plugins.go#L65-L106), mutating admission plugins run before the mutating webhooks.

**Comment by liggitt** [comment]
> I would consider a blocker for scheduling profiles being GA. But it would be nice to backport to 1.19 at least, when we graduated to beta.

I'm confused. This field existed with this default before the concept of scheduling profiles was added.

**Comment by liggitt** [comment]
Setting it in PrepareForCreate would mean pods created before the field was introduced would be returned with an empty field value and would be missed by the default scheduler

**Comment by alculquicondor** [comment]
> I'm confused. This field existed with this default before the concept of scheduling profiles was added.

That's true. But with scheduling profiles it becomes easier for users to have more profiles and thus there is more motivation to have automation (such as webhooks) around them. But since backporting poses extra risk, I'm ok with leaving this to 1.20.

> Setting it in PrepareForCreate would mean pods created before the field was introduced would be returned with an empty field value and would be missed by the default scheduler

That would only affect pods that didn't get scheduled, right? However, the field is there since 1.6

**Comment by liggitt** [comment]
> > Setting it in PrepareForCreate would mean pods created before the field was introduced would be returned with an empty field value and would be missed by the default scheduler
> 
> That would only affect pods that didn't get scheduled, right? However, the field is there since 1.6

in the first release this field was introduced, the intent was for existing pods, when read, to have the same `spec.schedulerName` value they would have if they were newly created without setting that field. Populating the field via API defaulting on deserialization is how we accomplish that.

Now that the field has existed for O(years), it is likely a safe assumption that all persisted pods have been written at least once since 1.6 and therefore have the field populated.

Moving the defaulting from API defaulting to the REST handler would help webhooks that wanted to distinguish between user-specified and defaulted `spec.schedulerName`, but would break mutating webhooks that currently expect to see the chosen `spec.schedulerName` and take action based on it.

We'd also need to populate `spec.schedulerName` in PrepareForUpdate to avoid breaking clients that update pods with that field unpopulated (would cause a validation error saying the field was immutable otherwise)

**Comment by alculquicondor** [comment]
> but would break mutating webhooks that currently expect to see the chosen spec.schedulerName and take action based on it.

Is it fair to do for 1.20 then? With a proper `ACTION REQUIRED` release note.

**Comment by liggitt** [comment]
> Is it fair to do for 1.20 then? With a proper `ACTION REQUIRED` release note.

Not really... API changes like that (removing a field from a GA type or changing an API-level default) are not something we change for the lifetime of a given API version. Admission webhooks today can rely on defaulting having been applied to the object they receive.

**Comment by lavalamp** [comment]
The problem basically is that a webhook can't tell if a user set this field, or it was defaulted.

Post 1.18, we could figure this out from looking at the SSA field managers if those were set prior to calling webhooks. @apelisse do you recall if we set field managers that early?

I think it's pretty reasonable to ask for SOME way of figuring this out.

Let's think about who this would break, if we move the defaulting. It's a webhook that is looking for `spec.schedulerName == "default"` and is going to set some alternative default under the assumption that the user didn't actually set the field specifically (which may be wrong). Now, if that webhook sees this field is unset, it will assume that's a deliberate choice by the user and not change the field. Then apiserver will default it. The result is, only pods which specifically request the default scheduler will get acted on by this webhook.

I think this is not a terrible failure mode (at least as compared to some others), and while it will be annoying for the webhook author to recompile, their webhook will now be able to tell the difference between unset and deliberately set, and will be able to behavior appropriately.

So, I think this is a strong candidate for stretching the rules.

**Comment by ahg-g** [comment]
It is indeed tricky. One **ugly** solution is to add an annotation indicating that a default value was applied when unset. This would preserve the current behavior, but allow web hooks that want to tell the difference between unset and deliberately set default.

**Comment by alculquicondor** [comment]
@liggitt any thoughts on the last 2 suggestions?

**Comment by liggitt** [comment]
> Let's think about who this would break, if we move the defaulting. It's a webhook that is looking for spec.schedulerName == "default" and is going to set some alternative default under the assumption that the user didn't actually set the field specifically (which may be wrong).

It would also break a webhook that assumed the spec.schedulerName field was present and non-empty in the received pod object

**Comment by liggitt** [comment]
> @liggitt any thoughts on the last 2 suggestions?

Adding annotations indicating "this particular field was defaulted" seems pretty ad-hoc and non-scalable.

If we dropped `omitempty` to make spec.schedulerName always be serialized, even when an empty string, that would avoid nil-pointer breakage for non-go webhooks. The scenario Daniel identified would still be affected, but I agree it is a fairly narrow scenario.

**Comment by ahg-g** [comment]
Adding annotations is indeed ad-hoc, but it is something worth considering in a situation where the other solution is breaking a commitment to a stable api. I agree though that the way we are breaking the api is limited and easy to adapt to.

Out of courtesy why adding annotations non-scalable?

**Comment by liggitt** [comment]
> Out of courtesy why adding annotations non-scalable?

I meant that paired, "ad-hoc and non-scalable", meaning we couldn't apply the same approach to all fields we default

**Comment by ahg-g** [comment]
s/courtesy/curiosity :)

right, wasn't suggesting to use it for all fields, but this is indeed ad-hoc if applied to a single field (again it is a tradeoff between this and how committed we are to stable apis)

**Comment by lavalamp** [comment]
I'm OK with that.

**Comment by apelisse** [comment]
Answering super late,  but still:
> Post 1.18, we could figure this out from looking at the SSA field managers if those were set prior to calling webhooks. @apelisse do you recall if we set field managers that early?

Yes, webhooks receive the managedFields that have just been set.

Also, it makes me a little sad that we're needing a mechanism that CRDs don't have. We still need a long-term solution to this problem.

**Comment by alculquicondor** [comment]
What is the suggestion with SSA? Is this some information that webhooks can use to tell if the apiserver set a field or it was done by the user?

**Comment by apelisse** [comment]
Yeah, but that will only work if the resource was server-side applied, not if it was "POST"ed (in that case, we unfortunately include the defaults), so I don't know if that's good enough.

**Comment by alculquicondor** [comment]
Most Pods would be created from Deployments or other workload APIs, so, no, it wouldn't be enough.

**Comment by alculquicondor** [comment]
ping @liggitt :)

can we proceed with
https://github.com/kubernetes/kubernetes/issues/93067#issuecomment-694300442 ?

**Comment by liggitt** [comment]
> We can move the defaulting to the REST handler and remove "omitempty" from the serialization.

For the impact to pod spec, this seems reasonable. 

Removing "omitempty" has the really unfortunate side-effect of changing the serialization for all types that embed pod template as well. When they don't specify a scheduler name (which most deployments/replicasets/... don't), they will all start serializing an empty string. Does that cause problems other than being really ugly?
